### PR TITLE
fix(vue-renderer): add the csp hash if `unsafe-inline` hasn't been specified

### PR DESCRIPTION
**Pull request by sambowler**
Only adds the CSP hash if `'unsafe-inline'` hasn't been specified.

## Types of changes

- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
If both a hash and `'unsafe-inline'` are present, browsers ignore the rule (see screenshot). This change will allow users to specify the option effectively.

Screen Shot 2019-03-26 at 10 11 18

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

File: packages/vue-renderer/src/renderer.js [hunk 1]
```diff
@@ -408,7 +408,11 @@ export default class VueRenderer {
 
     // Calculate CSP hashes
     const cspScriptSrcHashes = []
-    if (this.context.options.render.csp) {
+    const csp = this.context.options.render.csp
+    const containsUnsafeInlineScriptSrc = csp && csp.policies['script-src'] && csp.policies['script-src'].indexOf('\'unsafe-inline\'') !== -1
```
Comment (manniL) [inline packages/vue-renderer/src/renderer.js]:
```suggestion
    const containsUnsafeInlineScriptSrc = csp && csp.policies['script-src'] && csp.policies['script-src'].includes('\'unsafe-inline\'')
```

File: packages/vue-renderer/src/renderer.js [hunk 2]
```diff
@@ -408,7 +408,11 @@ export default class VueRenderer {
...
+
+    // Only add the hash if 'unsafe-inline' rule isn't present to avoid conflicts
```
Comment (manniL) [inline packages/vue-renderer/src/renderer.js]:
```suggestion
    // Only add the hash if 'unsafe-inline' rule isn't present to avoid conflicts (#5387)
```

**Review by TheAlexLichter — CHANGES_REQUESTED** [review]
Hey! Good catch. I left you a few suggestions.

File: packages/vue-renderer/src/renderer.js [hunk 3]
```diff
@@ -408,7 +408,11 @@ export default class VueRenderer {
 
     // Calculate CSP hashes
     const cspScriptSrcHashes = []
-    if (this.context.options.render.csp) {
+    const csp = this.context.options.render.csp
+    const containsUnsafeInlineScriptSrc = csp && csp.policies && csp.policies['script-src'] && csp.policies['script-src'].indexOf('\'unsafe-inline\'') !== -1
```
Comment (manniL) [inline packages/vue-renderer/src/renderer.js]:
```suggestion
    const containsUnsafeInlineScriptSrc = csp && csp.policies && csp.policies['script-src'] && csp.policies['script-src'].includes(`'unsafe-inline'`)
```

**Comment by sambowler** [comment]
Thanks @manniL, I've applied those suggestions. 🙂

File: packages/vue-renderer/src/renderer.js [hunk 4]
```diff
@@ -408,7 +408,11 @@ export default class VueRenderer {
 
     // Calculate CSP hashes
     const cspScriptSrcHashes = []
-    if (this.context.options.render.csp) {
+    const csp = this.context.options.render.csp
+    const containsUnsafeInlineScriptSrc = csp && csp.policies && csp.policies['script-src'] && csp.policies['script-src'].includes(`'unsafe-inline'`)
```
Comment (manniL) [inline packages/vue-renderer/src/renderer.js]: Maybe even 

```suggestion
    const containsUnsafeInlineScriptSrc = csp.policies && csp.policies['script-src'] && csp.policies['script-src'].includes(`'unsafe-inline'`)
```

as the csp truth check is already done before

Comment (sambowler) [reply packages/vue-renderer/src/renderer.js]: I think this is still necessary -- if `csp` were undefined but we tried to access `policies` we'd get a "Cannot read property..." error

Comment (manniL) [reply packages/vue-renderer/src/renderer.js]: Whoops, you are right!

**Comment by pi0** [comment]
@manniL Can you please review this again?

**Review by TheAlexLichter — APPROVED** [review]

**Comment by manniL** [comment]
@pi0 fine from my POV :relaxed:

**Comment by williamchong007** [comment]
Defining BOTH `unsafe-inline` and hash in CSP is actually a valid use case, it is for backward compatibility of CSPv1. That's why browsers with CSPv2+ support ignore it when hash is present at the same time.

[MDN reference](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src):

    See unsafe inline script for an example. Specifying nonce makes a modern browser ignore 'unsafe-inline' which could still be set for older browsers without nonce support.

It is also a recommended practice suggested by google
image

With this PR in, how can I specify BOTH `unsafe-inline` and SHA hash  for my CSP rule?

@manniL @sambowler

**Comment by manniL** [comment]
@williamchong007 Interesting. Please raise an extra issue so we can discuss it further there!

**Pull request merged**